File: marketing/create-video.py
# Video Creator - Python/MoviePy Version (Bulletproof)
# Install: pip install moviepy

from moviepy.editor import *
import glob
import os

print("🎬 Creating video...")

# Settings
audio_dir = "marketing/audio"
images_dir = "marketing/raw"
output_file = "marketing/processed/video-1-what-im-building.mp4"

# Get all audio files in order
audio_files = [
    f"{audio_dir}/01-hook.mp3",
    f"{audio_dir}/02-vision.mp3",
    f"{audio_dir}/03-treasury.mp3",
    f"{audio_dir}/04-giving.mp3",
    f"{audio_dir}/05-apps.mp3",
    f"{audio_dir}/06-automation.mp3",
    f"{audio_dir}/07-discord-cta.mp3",
    f"{audio_dir}/08-video-system.mp3",
    f"{audio_dir}/09-final-cta.mp3"
]

# Combine all audio
print("Merging audio tracks...")
audio_clips = [AudioFileClip(f) for f in audio_files]
final_audio = concatenate_audioclips(audio_clips)
total_duration = final_audio.duration

# Get all image files
image_files = sorted(glob.glob(f"{images_dir}/photo_*_2026-01-05_08-39-59.jpg"))
image_files.append(f"{images_dir}/photo_2026-01-05_08-42-31.jpg")

print(f"Found {len(image_files)} images")
print(f"Total audio duration: {total_duration:.1f} seconds")

# Calculate duration per image
duration_per_image = total_duration / len(image_files)

# Create video clips from images
print("Creating slideshow...")
video_clips = []
for img_path in image_files:
    clip = (ImageClip(img_path)
            .set_duration(duration_per_image)
            .resize(height=1920)  # Vertical video
            .on_color(size=(1080, 1920), color=(0,0,0), pos='center'))
    video_clips.append(clip)

# Concatenate all video clips
final_video = concatenate_videoclips(video_clips, method="compose")

# Add audio to video
final_video = final_video.set_audio(final_audio)

# Export
print("Exporting final video...")
final_video.write_videofile(
    output_file,
    fps=30,
    codec='libx264',
    audio_codec='aac',
    audio_bitrate='320k',
    preset='slow',
    ffmpeg_params=['-crf', '18']
)

# Cleanup
for clip in audio_clips + video_clips:
    clip.close()
final_audio.close()
final_video.close()

print(f"\n✅ Done! Video saved to: {output_file}")
print(f"Size: {os.path.getsize(output_file) / 1024 / 1024:.1f} MB")

File: marketing/create-video.js
const ffmpeg = require('fluent-ffmpeg');
const ffmpegPath = require('@ffmpeg-installer/ffmpeg').path;
const fs = require('fs');
const path = require('path');

ffmpeg.setFfmpegPath(ffmpegPath);

console.log('🎬 Creating video...\n');

const audioDir = 'audio';
const imagesDir = 'raw';
const outputFile = 'processed/video-1-what-im-building.mp4';

// Audio files in order
const audioFiles = [
    '01-hook.mp3', '02-vision.mp3', '03-treasury.mp3', '04-giving.mp3',
    '05-apps.mp3', '06-automation.mp3', '07-discord-cta.mp3',
    '08-video-system.mp3', '09-final-cta.mp3'
].map(f => path.join(audioDir, f));

// Get all image files
const imageFiles = [];
for (let i = 1; i <= 41; i++) {
    imageFiles.push(path.join(imagesDir, `photo_${i}_2026-01-05_08-39-59.jpg`));
}
imageFiles.push(path.join(imagesDir, 'photo_2026-01-05_08-42-31.jpg'));

console.log(`Found ${imageFiles.length} images`);
console.log(`Found ${audioFiles.length} audio files\n`);

// Step 1: Concatenate all audio files
console.log('Step 1: Merging audio tracks...');
const tempAudio = 'processed/temp-audio.mp3';

let audioConcat = ffmpeg();
audioFiles.forEach(file => audioConcat.input(file));

audioConcat
    .on('error', err => {
        console.error('Audio merge failed:', err.message);
        process.exit(1);
    })
    .on('end', () => {
        console.log('✓ Audio merged\n');

        // Step 2: Create slideshow with images
        console.log('Step 2: Creating slideshow from images...');

        // Calculate duration per image based on total audio length
        ffmpeg.ffprobe(tempAudio, (err, metadata) => {
            if (err) {
                console.error('Failed to get audio duration:', err.message);
                process.exit(1);
            }

            const totalDuration = metadata.format.duration;
            const durationPerImage = totalDuration / imageFiles.length;

            console.log(`Total duration: ${totalDuration.toFixed(1)}s`);
            console.log(`Duration per image: ${durationPerImage.toFixed(1)}s\n`);

            // Create slideshow
            let slideshow = ffmpeg();

            imageFiles.forEach(img => {
                slideshow.input(img).inputOptions([`-loop 1`, `-t ${durationPerImage}`]);
            });

            const filterComplex = imageFiles.map((_, i) =>
                `[${i}:v]scale=1080:1920:force_original_aspect_ratio=decrease,pad=1080:1920:(ow-iw)/2:(oh-ih)/2,setsar=1,fps=30[v${i}]`
            ).join(';');

            const concat = imageFiles.map((_, i) => `[v${i}]`).join('') + `concat=n=${imageFiles.length}:v=1:a=0[outv]`;

            slideshow
                .input(tempAudio)
                .complexFilter(`${filterComplex};${concat}`)
                .outputOptions([
                    '-map [outv]',
                    '-map ' + imageFiles.length + ':a',
                    '-c:v libx264',
                    '-preset slow',
                    '-crf 18',
                    '-pix_fmt yuv420p',
                    '-c:a aac',
                    '-b:a 320k',
                    '-shortest',
                    '-movflags +faststart'
                ])
                .output(outputFile)
                .on('progress', progress => {
                    process.stdout.write(`\rProgress: ${progress.percent ? progress.percent.toFixed(1) : 0}%`);
                })
                .on('error', err => {
                    console.error('\n\nVideo creation failed:', err.message);
                    fs.unlinkSync(tempAudio);
                    process.exit(1);
                })
                .on('end', () => {
                    console.log('\n\n✅ SUCCESS!');

                    const stats = fs.statSync(outputFile);
                    console.log(`\nFile: ${outputFile}`);
                    console.log(`Size: ${(stats.size / 1024 / 1024).toFixed(2)} MB`);
                    console.log(`Format: 1080x1920 vertical, H.264, 30fps`);
                    console.log('\nReady to upload to YouTube Shorts, TikTok, X!');

                    // Cleanup
                    fs.unlinkSync(tempAudio);
                })
                .run();
        });
    })
    .mergeToFile(tempAudio);
